runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,323 @@
+    if not bufnr then
+      return
+    end
+    local bufstate = assert(bufstates[bufnr])
+
+    ---@type integer
+    local client_id = args.data.client_id
+    bufstate.client_hilights[client_id] = nil
+    if vim.tbl_isempty(bufstate.client_hilights) then
+      bufstates[bufnr] = nil
+    end
+  end,
+})
+
+---|lsp-handler| for the method `textDocument/documentHighlight`
+---Store hints for a specific buffer and client
+---@param result? lsp.DocumentHighlight[]
+---@param ctx lsp.HandlerContext
+function M.on_document_highlight(err, result, ctx)
+  if err then
+    log.error('document highlight', err)
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  local client_id = assert(ctx.client_id)
+  local client_hilights = bufstate.client_hilights
+
+  ---@type table<integer, lsp.DocumentHighlight[]?>
+  local row_highlights = {}
+  for _, highlight in pairs(result or {}) do
+    for row = highlight.range['start'].line, highlight.range['end'].line do
+      local highlights = row_highlights[row] or {}
+      highlights[#highlights + 1] = highlight
+      row_highlights[row] = highlights
+    end
+  end
+  client_hilights[client_id] = row_highlights
+  api.nvim__redraw({ buf = bufnr, valid = true })
+end
+
+---@param bufnr integer
+local function refresh(bufnr)
+  local bufstate = assert(bufstates[bufnr])
+  local enabled = bufstate.enabled
+  if enabled == nil then
+    enabled = globalstate.enabled
+  end
+
+  if not enabled then
+    api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
+    api.nvim__redraw({ buf = bufnr, valid = true })
+    return
+  end
+
+  local params = util.make_position_params()
+  vim.lsp.buf_request(bufnr, ms.textDocument_documentHighlight, params)
+end
+
+local function debunce(f)
+  ---@type uv.uv_timer_t?
+  local timer = nil
+  return function(...)
+    local args = { ... }
+    if timer then
+      vim.uv.timer_stop(timer)
+      timer:close()
+      timer = nil
+    end
+    timer = assert(vim.uv.new_timer())
+    vim.uv.timer_start(
+      timer,
+      100,
+      0,
+      vim.schedule_wrap(function()
+        timer:close()
+        timer = nil
+        f(unpack(args))
+      end)
+    )
+  end
+end
+
+local debounced_refresh = debunce(refresh)
+
+api.nvim_create_autocmd('CursorMoved', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    if bufstates[bufnr] then
+      debounced_refresh(bufnr)
+    end
+  end,
+})
+api.nvim_create_autocmd('LspNotify', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+    if
+      client.supports_method(ms.textDocument_documentHighlight, { bufnr = bufnr })
+      and (
+        args.data.method == ms.textDocument_didChange
+        or args.data.method == ms.textDocument_didOpen
+      )
+    then
+      debounced_refresh(bufnr)
+    end
+  end,
+})
```

Comment:
this will have the side effect that requests triggered by other plugins or custom user functions would also trigger a new refresh - I'm not sure if that's desired?

Why is this additional LspNotify autocmd necessary?